### ISSUE TITLE
Make it possible to run the dashboard behind a reverse proxy or load balancer

### ISSUE THREAD

**Issue by fiotti** [dotnet/aspire#5134]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I would like to be able to host Aspire Dashboard behind a reverse proxy such as NGINX. Unfortunately, Aspire Dashboard doesn't take `X-Forwarded-*` headers into consideration, and doesn't allow to configure a path base.

As an example, assume Aspire Dashboard is accessible internally from http://aspire-dashboard-service:18888/ and externally from https://example.com/dev/aspire-dashboard/ through a reverse proxy.

The issue of the path base has already been described in #4528 and #4159. Additionally, authentication through OpenID Connect fails to set the correct value for `redirect_uri`, as it uses the internal schema (HTTP) rather than the external one (HTTPS).

### Describe the solution you'd like

It's possible to add support for reverse proxies and load balancers as described here:
https://learn.microsoft.com/aspnet/core/host-and-deploy/proxy-load-balancer

Ideally, all the featured described in previous page (except for certificate forwarding) should be configurable by the user, namely:
- Forwarded headers from the reverse proxy or load balancer
- Header names, known proxies and forward limit
- Forcing the assumption that external address is always HTTPS (necessary when the reverse proxy does not support `X-Forwarded-*` headers)
- Path base

**Comment by ewassef** [comment]
I created a PR for this

**Comment by zheyi-zhu** [comment]
Still no supported i think as of now. 

My prod setup DNS => Azure Frontdoor (SSL Termination) => ALB Controller => Gateway API (80) not working.
Dev setup DNS => ALB Controller => Gateway API (SSL termination) works.

**Comment by davidfowl** [comment]
We fixed this in 9.5 https://learn.microsoft.com/en-us/dotnet/aspire/whats-new/dotnet-aspire-9.5#reverse-proxy-support

**Comment by zheyi-zhu** [comment]
is it released? I just pulled the image and it does not work

**Comment by davidfowl** [comment]
Use the nightly package, there's no tagged release for 9.5 as yet

cc @joperezr

**Comment by zheyi-zhu** [comment]
Thanks for the work. Will wait for the release.

**Comment by joperezr** [comment]
It actually shipped earlier today `mcr.microsoft.com/dotnet/aspire-dashboard:9.5`

**Comment by zheyi-zhu** [comment]
Yep I pulled 9.5: `mcr.microsoft.com/dotnet/aspire-dashboard:9.5` 

ConfigMap:
```yaml
ASPIRE_DASHBOARD_FORWARDEDHEADERS_ENABLED: "true"

```

Log: notice the `redirect_uri` is still http.  
```
dbug: Aspire.Dashboard.Authentication.FrontendCompositeAuthenticationHandler[9]
      AuthenticationScheme: FrontendComposite was not authenticated.
dbug: Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler[53]
      HandleChallenge with Location: https://login.microsoftonline.com/<tenant-id>/oauth2/authorize?client_id=<client-id>&
      redirect_uri=http%3A%2F%2F<my-domain>%2F&
      response_type=code&
      scope=openid%20profile&
      code_challenge=<code-challenge>&
      code_challenge_method=S256&
      response_mode=form_post&
      nonce=<nonce>&
      state=<state>&
      x-client-SKU=ID_NET8_0&
      x-client-ver=7.1.2.0; and Set-Cookie: .AspNetCore.OpenIdConnect.Nonce.<nonce-cookie>; expires=Tue, 14 Oct 2025 22:36:42 GMT; path=/; secure; samesite=none; httponly,.AspNetCore.Correlation.<correlation>; expires=Tue, 14 Oct 2025 22:36:42 GMT; path=/; secure; samesite=none; httponly.
info: Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler[12]
      AuthenticationScheme: OpenIdConnect was challenged.
```

Maybe it is related to my setup, but I have other dozens of apps setup in the same way, they are working fine.

I think this is the [PR-10388](https://github.com/dotnet/aspire/pull/10388): im not very good at navigating in github so im not sure if this is released or not, although it is in the milestone.

**Comment by davidfowl** [comment]
@zheyi-zhu can you confirm the env variable is getting set appropriately in the container?

**Comment by zheyi-zhu** [comment]
TLDR: The `X-Forwarded-Proto=https` is not present in my setup.

Details:
My production setup here is 

DNS => Frontdoor origin (SSL Termination) => Azure Kubernetes ALB Controller => Gateway (80) => HttpRoute to aspire-dashboard service to the pods. 

I added some debug code in the repo, build it and then logs and find out we only have 3 x- standard headers: 
* `X-Original-Proto = http`
* `X-Original-Host = my-host` 
* `x-forwarded-for = some-ip`

I have to endup inject a `X-Forwarded-Proto=https` to make it work in my production setup.

Since I already build the image, so would just use the image I temporarily built.

---

I want to ask when we can support `Authentication__AzureAd__RedirectUri` so that I can directly set it with https

**Comment by davidfowl** [comment]
@zheyi-zhu did you verify that this code path was getting hit https://github.com/dotnet/aspire/blob/911e1f62c0caa3fae683218f24ef0cd0be3e1efb/src/Aspire.Dashboard/DashboardWebApplication.cs#L243

**Comment by zheyi-zhu** [comment]
@davidfowl yes. it was hit. The code works the way it is.

**Comment by fiotti** [comment]
If I may interject and share my opinion, a single boolean such as `ASPIRE_DASHBOARD_FORWARDEDHEADERS_ENABLED` is not sufficient as a generic configuration for services behind all kinds of reverse-proxies. The ["Configure ASP.NET Core to work with proxy servers and load balancers"](https://learn.microsoft.com/aspnet/core/host-and-deploy/proxy-load-balancer) page from Microsoft's official ASP.NET Core documentation explains this problem in detail.

We ended up exposing the following options in the ASP.NET Core services developed by the company I work for, I would recommend doing something similar:
```jsonc
{
  "Hosting": {
    "Hosts": {
      "example-local-dev": {
        "Host": "localhost:*",
        "PathBase": "/"
      },
      "example-nginx-proxy": {
        "Host": "example.com",
        "PathBase": "/dev",
        "Scheme": "https" /* Necessary when the reverse-proxy does not support "X-Forward-Proto", otherwise null. */
      },
      "fallback": {
        "Host": "*",
        "PathBase": "/"
      }
    }
  },
  "ForwardedHeaders": {
    /* Same as "Microsoft.AspNetCore.Builder.MicrosoftForwardedHeadersOptions". */
    "ForwardedForHeaderName": "X-Forwarded-For",
    "ForwardedHostHeaderName": "X-Forwarded-Host",
    "ForwardedProtoHeaderName": "X-Forwarded-Proto",
    "ForwardedPrefixHeaderName": "X-Forwarded-Prefix",
    "OriginalForHeaderName": "X-Original-For",
    "OriginalHostHeaderName": "X-Original-Host",
    "OriginalProtoHeaderName": "X-Original-Proto",
    "OriginalPrefixHeaderName": "X-Original-Prefix",
    "ForwardedHeaders": [
      /* Same as "Microsoft.AspNetCore.HttpOverrides.ForwardedHeaders", or "*" for all. */
      "X-Forwarded-For",
      "X-Forwarded-Host",
      "X-Forwarded-Proto"
    ],
    "ForwardLimit": 1, /* Number of reverse-proxies in the chain. */
    "KnownProxies": [
      "::1"
    ],
    "KnownNetworks": [
      "127.0.0.1/8"
    ],
    "AllowedHosts": [
      "localhost",
      "example.com"
    ],
    "RequireHeaderSymmetry": false
  },
  "CertificateForwarding": {
    /* The Aspire Dashboard does not currently make use of client certificates. */
    "Enabled": false,
    "CertificateHeader": "X-Client-Cert"
  }
}
```

[JSON-Schema](https://alad00.github.io/JSON-Schemas/Alad/appsettings/definitions/Alad.Utility.ProxyLoadBalancer.json) for reference (in Italian).

**Comment by WeihanLi** [comment]
Wondering why we don't prefer the `ASPNETCORE_FORWARDEDHEADERS_ENABLED` instead?

**Comment by davidfowl** [comment]
Will leave this issue opened.

**Comment by zheyi-zhu** [comment]
@WeihanLi I think it is heavily relying on all the connected dots between the client and the final destination services forward the required headers. If one actor that does not forward the headers properly, this method would fail. 

In my case, the ssl termination happens at Azure Frontdoor. It then forwards the traffic in http to Azure Kubernetes Service, assuming all traffic is now http. The assumption is correct but it removes the `X-Forwarded-Proto` header from my observation.

`ASPNETCORE_FORWARDEDHEADERS_ENABLED` is nice and all but does not cover all the use cases.

**Comment by ckraenzel** [comment]
@davidfowl Is there a way to set ASPIRE_DASHBOARD_FORWARDEDHEADERS_ENABLED without deploying the dashboard as standalone ?

**Comment by davidfowl** [comment]
What do you mean? Set it where?

**Comment by ckraenzel** [comment]
We're making our first steps with aspire, we get our apps running behind app gateway as reverse proxy, however the dashboard that is set up is not accessible, redirect after authentication fails. Do we need to deploy the dashboard standalone or is there a way to enable correct use of forward headers ?

**Comment by davidfowl** [comment]
So you are talking about the dashboard hosting in Azure container apps? No you don't control that, we can talk to the ACA team about it.

**Comment by tpflueger** [comment]
I've ran into an issue that might be similar to this. I have deployed the standalone aspire dashboard to AWS using Application Load Balancer that terminates SSL. When I try to navigate to the hosted domain like https://aspire.*.net, I get redirected to https://aspire.*.net/structuredlogs with a 404.

However, if I go to https://aspire.*.net/login I get directed to the login page that I can enter the token. If I enter the token, I get redirected to https://aspire.*.net/structuredlogs with a 404.

**Comment by snakex64** [comment]
After 2 years, do we have an ETA on this ? 🥲
We are hosting aspire in our QA environments and we can't create sub domains for each of them. Going through a reverse proxy with an added query path like `/aspire` is the only way

**Comment by davidfowl** [comment]
We accept PRs! More importantly, there are enough users on this thread with scenarios it should be enough requirements to verify any solution proposed.

https://github.com/microsoft/aspire/issues/5134#issuecomment-3406039102 might be the most thorough.

**Comment by DiscoPYF** [comment]
We also have this issue hosting the standalone Aspire dashboard behind nginx.
`https://demo-myproduct.company.com/otel` -> `http://aspire-dashboard:18888`

The dashboard is configured to use OIDC with Keycloak as provider, in a Docker Compose deployment. This is for a demo instance of our product. We use the Aspire dashboard because the visualization is much simpler than Graphana.

We are hosting it on a separate port as a workaround: `https://demo-myproduct.company.com:18888`

**Comment by DiscoPYF** [comment]
@davidfowl In reference to your comment on https://github.com/microsoft/aspire/pull/6963#issuecomment-3782690368

> Originally, we though this would be a simple change but it turned out to make it really complete was a yak shaving exercise and we lacked the right testing infrastructure to find all of the places that needed to change to make this work for realz. I will close this PR but we can take another stab at this now that we have a bit more experience with the problems and have ways to verify it.

Is this issue still depending on setting up the right testing infrastructure first? Do you have some recommendation for someone willing to take another stab at this? 👀

**Comment by pasisavolainen** [comment]
Could we get at least the correct BasePath? I'm very happy for you all that you can get hostnames and certificates as you like in your environments but some of us do live under regime and admin on hostname locked IIS is the extent of our rights. We have to run whatever we want from virtual paths.

**Comment by fischetm** [comment]
Looks like: aspire blazor code change or run an nginx with url response rewrite - or iis if you have. I will go nginx in docker sidecar to the aspire dashboard standalone docker

**Comment by pasisavolainen** [comment]
Just url/html/json rewriting with proxy did not work. Some url chunks are passed through websocket (blazor?) and handling those would involve decoding and manipulating those.
